saito-core/src/core/util/rate_limiter.rs

```diff
@@ -0,0 +1,47 @@
+use std::collections::HashMap;
+
+
+#[derive(Debug, Clone)]
+pub struct RateLimiter {
+    limits: HashMap<String, usize>,
+    windows: HashMap<String, u64>, // Use u64 for duration in milliseconds
+    request_counts: HashMap<String, usize>,
+    last_request_times: HashMap<String, u64>, // Use u64 for timestamps
+}
+
+impl RateLimiter {
+    pub fn new() -> Self {
+        RateLimiter {
+            limits: HashMap::new(),
+            windows: HashMap::new(),
+            request_counts: HashMap::new(),
+            last_request_times: HashMap::new(),
+        }
+    }
+
+    pub fn set_limit(&mut self, action: &str, limit: usize, window: u64) {
+        self.limits.insert(action.to_string(), limit);
```

Comment:
since we are inserting and reading all 4 maps all the time, can't we use a single HashMap<String, SomeData> for this?